include/nbl/video/IGPUCommandBuffer.h

```diff
@@ -227,6 +232,93 @@ class IGPUCommandBuffer :
+        Cmd* cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
+        if (!cmd)
+        {
+            void* nextSegmentMem = m_cmdpool->m_commandSegmentPool.allocate(IGPUCommandPool::COMMAND_SEGMENT_SIZE, alignof(IGPUCommandPool::CommandSegment));
+            if (nextSegmentMem == nullptr)
+                return nullptr;
+
+            IGPUCommandPool::CommandSegment* nextSegment = new (nextSegmentMem) IGPUCommandPool::CommandSegment;
+
+            // cmd = m_segmentListTail->allocate<Cmd, Args...>(args...);
+            cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
+            if (!cmd)
+                return nullptr;
+
+            m_segmentListTail->params.m_next = nextSegment;
+            m_segmentListTail = m_segmentListTail->params.m_next;
+        }
+
+        if (m_segmentListHeadItr.m_cmd == nullptr)
+            m_segmentListHeadItr.m_cmd = cmd;
+
+        return cmd;
+    }
+
+    IGPUCommandPool::CommandSegment::Iterator m_segmentListHeadItr = {};
+    IGPUCommandPool::CommandSegment* m_segmentListTail = nullptr;
+
+    private:
+        void deleteCommandSegmentList()
+        {
+            IGPUCommandPool::CommandSegment::Iterator itr = m_segmentListHeadItr;
+
+            if (itr.m_segment && itr.m_cmd)
+            {
+                bool lastCmd = itr.m_cmd->m_size == 0u;
+                while (!lastCmd)
+                {
+                    IGPUCommandPool::ICommand* currCmd = itr.m_cmd;
+                    IGPUCommandPool::CommandSegment* currSegment = itr.m_segment;
+
+                    itr.m_cmd = reinterpret_cast<IGPUCommandPool::ICommand*>(reinterpret_cast<uint8_t*>(itr.m_cmd) + currCmd->m_size);
+                    currCmd->~ICommand();
+                    // No need to deallocate currCmd because it has been allocated from the LinearAddressAllocator where deallocate is a No-OP and the memory will
+                    // get reclaimed in ~LinearAddressAllocator
+
+                    if ((reinterpret_cast<uint8_t*>(itr.m_cmd) - reinterpret_cast<uint8_t*>(itr.m_segment)) > IGPUCommandPool::CommandSegment::STORAGE_SIZE)
+                    {
+                        IGPUCommandPool::CommandSegment* nextSegment = currSegment->params.m_next;
+                        if (!nextSegment)
+                            break;
+
+                        currSegment->~CommandSegment();
+                        m_cmdpool->m_commandSegmentPool.deallocate(currSegment, IGPUCommandPool::COMMAND_SEGMENT_SIZE);
+
+                        itr.m_segment = nextSegment;
+                        itr.m_cmd = reinterpret_cast<IGPUCommandPool::ICommand*>(itr.m_segment->m_data);
+                    }
+
+                    lastCmd = itr.m_cmd->m_size == 0u;
+                    if (lastCmd)
+                    {
+                        currSegment->~CommandSegment();
+                        m_cmdpool->m_commandSegmentPool.deallocate(currSegment, IGPUCommandPool::COMMAND_SEGMENT_SIZE);
```

Comment:
you need to check if you can do that.

Reply:
Check what exactly: destruction of `CommandSegment` or its deallocation? AFAICT we should always be able to destroy the `currSegment`.

Reply:
I meant the deallocation

Reply:
Resolved.